File: xr18/lib/src/main/java/com/mixer/xr18/lib/presentation/MixerStateFormatter.java
package com.mixer.xr18.lib.presentation;

import com.mixer.xr18.lib.domain.model.ChannelState;
import com.mixer.xr18.lib.domain.model.EqBands;
import com.mixer.xr18.lib.domain.model.MixerState;

/**
 * Converts raw [MixerState] to human-readable natural text.
 * Used for log output, status display, or debug UIs.
 */
public class MixerStateFormatter {
    private static final String TAG = "MixerFormatter";

    /**
     * Render the full mixer state as a natural-language block.
     */
    public static String format(MixerState state) {
        StringBuilder sb = new StringBuilder();

        sb.append("══════════════════════════════════════=\n");
        sb.append("  XR18 Mixer Status\n");
        sb.append("═══════════════════════════════════════\n");

        if (state.device != null) {
            sb.append("\n┌ Device Info ───────────────────────┐\n");
            sb.append("│ Name:        ").append(state.device.name).append("\n");
            sb.append("│ Model:       ").append(state.device.model).append("\n");
            sb.append("│ IP:          ").append(state.device.ipAddress).append("\n");
            sb.append("│ Firmware:    ").append(state.device.firmwareVersion).append("\n");
            sb.append("│ Port:        ").append(state.device.port).append("\n");
            sb.append("└──────────────────────────────────────┘\n");
        }

        sb.append("\n┌ Channel Status ─────────────────────┐\n");
        for (ChannelState ch : state.channels) {
            sb.append(formatChannel(ch)).append("\n");
        }
        sb.append("└──────────────────────────────────────┘\n");

        return sb.toString();
    }

    public static String formatChannel(ChannelState ch) {
        String num = String.format("%02d", ch.channelNumber);
        String fPct = ch.faderPercent();
        String fDb = ch.faderDbString();
        String mute = ch.muted ? "ON" : "OFF";
        String pan = ch.panString();

        String line1 = "  CH" + num + " Fader: " + fPct + " (" + fDb + "), Mute: " + mute + ", Pan: " + pan;

        String preampLine = String.format("Preamp: %.1f dB", ch.preampGain);
        String eqLine = ch.eqEnabled
            ? "EQ: ON [" + formatEqBands(ch.eqBands) + "]"
            : "EQ: OFF";

        if (ch.eqEnabled || ch.preampGain != 0f) {
            return line1 + "\n       " + preampLine + ", " + eqLine;
        }
        return line1;
    }

    private static String formatEqBands(EqBands bands) {
        return String.format("1:%.1fdB 2:%.1fdB 3:%.1fdB 4:%.1fdB",
            bands.band1, bands.band2, bands.band3, bands.band4);
    }

    /**
     * Short one-line summary of all channels.
     */
    public static String formatCompact(MixerState state) {
        StringBuilder sb = new StringBuilder();
        for (ChannelState ch : state.channels) {
            String n = String.format("%02d", ch.channelNumber);
            String m = ch.muted ? "M" : ".";
            String f = String.format("%.0f", ch.fader * 100);
            sb.append("[").append(n).append(":").append(f).append("%").append(m).append("]");
        }
        return sb.toString();
    }
}

File: xr18/lib/src/main/java/com/mixer/xr18/lib/presentation/MixerStateFormatter.kt
package com.mixer.xr18.lib.presentation

import android.util.Log
import com.mixer.xr18.lib.domain.model.ChannelState
import com.mixer.xr18.lib.domain.model.MixerState

/**
 * Converts raw [MixerState] to human-readable natural text.
 * Used for log output, status display, or debug UIs.
 */
object MixerStateFormatter {

    private const val TAG = "MixerFormatter"

    /**
     * Render the full mixer state as a natural-language block.
     * Shows device info + all 16 channel statuses.
     */
    fun format(state: MixerState): String {
        val sb = StringBuilder()

        // Header
        sb.appendLine("═══════════════════════════════════════")
        sb.appendLine("  XR18 Mixer Status")
        sb.appendLine("═══════════════════════════════════════")

        state.device?.let { dev ->
            sb.appendLine("")
            sb.appendLine("┌ Device Info ───────────────────────┐")
            sb.appendLine("│ Name:        ${dev.name}")
            sb.appendLine("│ Model:       ${dev.model}")
            sb.appendLine("│ IP:          ${dev.ipAddress}")
            sb.appendLine("│ Firmware:    ${dev.firmwareVersion}")
            sb.appendLine("│ Port:        ${dev.port}")
            sb.appendLine("└──────────────────────────────────────┘")
        }

        sb.appendLine("")
        sb.appendLine("┌ Channel Status ─────────────────────┐")

        state.channels.forEach { ch ->
            sb.appendLine(formatChannel(ch))
        }

        sb.appendLine("└──────────────────────────────────────┘")
        return sb.toString()
    }

    /**
     * Format a single channel as a readable one-liner.
     *
     * Example output:
     *   CH01 Fader: 60.6% (-8.2 dB), Mute: OFF, Pan: C
     *   CH02 Fader: 75.0% (-3.2 dB), Mute: ON,  Pan: L40
     */
    fun formatChannel(ch: ChannelState): String {
        val num  = "%02d".format(ch.channelNumber)
        val fPct = ch.faderPercent()
        val fDb  = ch.faderDbString()
        val mute = if (ch.muted) "ON"  else "OFF"
        val pan  = ch.panString()

        val faderLine = "Fader: $fPct ($fDb)"
        val muteLine  = "Mute: $mute"
        val panLine   = "Pan: $pan"

        val line1 = "  CH$num $faderLine, $muteLine, $panLine"

        // Preamp + EQ detail on second line (only if non-zero or active)
        val preampLine = "Preamp: ${"%.1f dB".format(ch.preampGain)}"
        val eqLine = if (ch.eqEnabled) {
            "EQ: ON [${formatEqBands(ch.eqBands)}]"
        } else {
            "EQ: OFF"
        }

        val line2 = if (ch.eqEnabled || ch.preampGain != 0f) {
            "       $preampLine, $eqLine"
        } else {
            ""
        }

        return if (line2.isNotEmpty()) "$line1\n$line2" else line1
    }

    private fun formatEqBands(bands: com.mixer.xr18.lib.domain.model.EqBands): String {
        return listOf(bands.band1, bands.band2, bands.band3, bands.band4)
            .mapIndexed { i, v -> "${i + 1}:${"%.1f".format(v)}dB" }
            .joinToString(" ")
    }

    /**
     * Short one-line summary of all channels — useful for a compact log line.
     */
    fun formatCompact(state: MixerState): String {
        val parts = state.channels.map { ch ->
            val n  = "%02d".format(ch.channelNumber)
            val m  = if (ch.muted) "M" else "."
            val f  = "%.0f".format(ch.fader * 100)
            "[$n:$f%$m]"
        }
        return parts.joinToString(" ")
    }

    /**
     * Write formatted state to android.util.Log at INFO level.
     */
    fun log(state: MixerState) {
        Log.i(TAG, format(state))
    }
}
